c++/src/kj/async-inl.h

```diff
@@ -1470,4 +1474,413 @@ PromiseForResult<Func, void> Executor::executeAsync(Func&& func) const {
+// these return_* functions live in a CRTP mixin.
+
+static stdcoro::coroutine_handle<void> noopCoroutine() noexcept;
+// TODO(msvc): Coroutines TS implementation doesn't have std::experimental::noop_coroutine(), so
+//   we'll roll our own below. This function is static so we can keep libkj coroutines header-only.
+
+template <typename T>
+template <typename U>
+class Coroutine<T>::Awaiter {
+  // Wrapper around a co_await'ed promise and some storage space for the result of that promise.
+  // The compiler arranges to call our await_suspend() to suspend, and we forward the call to the
+  // enclosing coroutine's promise adapter, which arranges to be woken up when the awaited promise
+  // is settled. Once that happens, the enclosing coroutine's promise adapter resumes the coroutine,
+  // which transitively calls await_resume() to unwrap the awaited promise result.
+
+public:
+  explicit Awaiter(Promise<U> promise): promise(kj::mv(promise)) {}
+  Awaiter(Awaiter&&) = default;
+  ~Awaiter() noexcept(false) {
+    unwindDetector.catchExceptionsIfUnwinding([this]() {
+      // No need to check for a moved-from state, promise will just ignore the nullification.
+      promise = nullptr;
+    });
+  }
+  KJ_DISALLOW_COPY(Awaiter);
+
+  bool await_ready() const { return false; }
+  // This could return "`promise.node.get()` is safe to call" instead, which would make suspension-
+  // less co_awaits possible for immediately-fulfilled promises. However, we need an Event to figure
+  // that out, and we won't have access to the Coroutine Event until await_suspend() is called. So,
+  // we must return false here. But read on, because await_suspend() has a trick up its sleeve to
+  // enable suspension-less co_awaits!
+  //
+  // We could also solve this by giving Awaiters a dummy Event that get immediately disarmed, or by
+  // adding a new isReady() virtual function to PromiseNode which everyone would have to implement.
+
+  U await_resume() {
+    auto& node = PromiseNode::from(promise);
+    node.get(result);
+
+    KJ_IF_MAYBE(exception, result.exception) {
+      kj::throwFatalException(kj::mv(*exception));
+    }
+
+    auto value = kj::_::readMaybe(result.value);
+    KJ_IASSERT(value != nullptr, "Neither exception nor value present.");
+    return U(kj::mv(*value));
+  }
+
+  stdcoro::coroutine_handle<void> await_suspend(Coroutine::Handle coroutine) {
+    auto& node = PromiseNode::from(promise);
+    auto& coroutineEvent = coroutine.promise();
+
+    node.onReady(&coroutineEvent);
+
+    if (coroutineEvent.isArmed()) {
+      // The result is immediately ready! Let's cancel our event.
+      coroutineEvent.disarm();
+
+      // We can resume ourselves by returning our own handle. This accomplishes the same thing as if
+      // we had extracted returned true from await_ready().
+      return coroutine;
+    } else {
+      // Otherwise, we must suspend, so we return a dummy coroutine_handle.
+      return noopCoroutine();
```

Comment:
Reading the docs, it looks like you could instead define `await_suspend` to return `bool`, and in this case `return true` to return control to the place where the coroutine was last resumed. (Whereas the other branch can `return false` to resume the coroutine immediately.)

Then I guess you don't need `noopCoroutine()`?

Reply:
Fixed: https://github.com/capnproto/capnproto/pull/1104/commits/3cd44f37fb74b71014a7ba994577439679f95947